cognite/extractorutils/uploader.py

```diff
@@ -380,19 +384,58 @@ def upload(self) -> None:
         max_delay=RETRY_MAX_DELAY,
         backoff=RETRY_BACKOFF_FACTOR,
     )
-    def _upload_batch(self, upload_this: List[Dict]) -> List[Dict]:
+    def _upload_batch(self, upload_this: List[Dict], retries=5) -> List[Dict]:
         if len(upload_this) == 0:
             return upload_this
 
         try:
             self.cdf_client.datapoints.insert_multiple(upload_this)
 
         except CogniteNotFoundError as ex:
-            self.logger.error("Could not upload data points to %s: %s", str(ex.not_found), str(ex))
+            if not retries:
+                raise ex
+
+            if not self.create_missing:
+                self.logger.error("Could not upload data points to %s: %s", str(ex.not_found), str(ex))
 
             # Get IDs of time series that exists, but failed because of the non-existing time series
             retry_these = [EitherId(**id_dict) for id_dict in ex.failed if id_dict not in ex.not_found]
 
+            if self.create_missing:
+                # Get the time series that can be created
+                create_these = [id_dict["externalId"] for id_dict in ex.not_found if "externalId" in id_dict]
+                self.cdf_client.time_series.create([TimeSeries(external_id=i) for i in create_these])
+
+                self.logger.info(f"Creating {len(create_these)} time series")
+
+                # Wait for the TS to be created before inserting (account for eventual consistency)
+                it = 0
+                while len(
+                    self.cdf_client.time_series.retrieve_multiple(external_ids=create_these, ignore_unknown_ids=True)
```

Comment:
Although I think create+retrieve by id is guaranteed to be strongly consistent (they query the SQL database directly), it's only the search and filter endpoints that are eventually consistent (they use Elasticsearch).

Reply:
Wow, TIL.

Confirmed it with the time lords team now. I'll remove this entire mess then.